exporter/cortexexporter/cortex.go

```diff
@@ -0,0 +1,354 @@
+		return fmt.Errorf("invalid metric type: wants summary points")
+	}
+
+	for _, pt := range metric.SummaryDataPoints {
+
+		time := int64(pt.GetTimeUnixNano())
+		mType := metric.GetMetricDescriptor().GetType()
+
+		// sum and count of the Summary should append suffix to baseName
+		baseName := getPromMetricName(metric.GetMetricDescriptor(), ce.namespace)
+
+		// treat sum as sample in an individual TimeSeries
+		sum := &prompb.Sample{
+			Value:     pt.GetSum(),
+			Timestamp: time,
+		}
+		sumLbs := createLabelSet(pt.GetLabels(), nameStr, baseName+sumStr)
+		addSample(tsMap, sum, sumLbs, mType)
+
+		// treat count as a sample in an individual TimeSeries
+		count := &prompb.Sample{
+			Value:     float64(pt.GetCount()),
+			Timestamp: time,
+		}
+		countLbs := createLabelSet(pt.GetLabels(), nameStr, baseName+countStr)
+		addSample(tsMap, count, countLbs, mType)
+
+		// process each percentile/quantile
+		for _, qt := range pt.GetPercentileValues() {
+			quantile := &prompb.Sample{
+				Value:     float64(qt.Value),
+				Timestamp: time,
+			}
+			percentileStr := strconv.FormatFloat(qt.Percentile, 'f', -1, 64)
+			qtLbs := createLabelSet(pt.GetLabels(), nameStr, baseName, quantileStr, percentileStr)
+			addSample(tsMap, quantile, qtLbs, mType)
+		}
+	}
+	return nil
+}
+
+// Because we are adhering closely to the Remote Write API, we must Export a
+// Snappy-compressed WriteRequest instance of the TimeSeries Metrics in order
+// for the Remote Write Endpoint to properly receive our Metrics data.
+func (ce *cortexExporter) export(ctx context.Context, TsMap map[string]*prompb.TimeSeries) error {
+	//Calls the helper function to convert the TsMap to the desired format
+	req, err := wrapTimeSeries(TsMap)
+	if err != nil {
+		return err
+	}
+
+	//Uses proto.Marshal to convert the WriteRequest into wire format (bytes array)
+	data, err := proto.Marshal(req)
+	if err != nil {
+		return err
+	}
+
+	//Makes use of the snappy compressor, as we are emulating the Remote Write package
+	compressedData := snappy.Encode(nil, data)
+
+	//Create the HTTP POST request to send to the endpoint
+	httpReq, err := http.NewRequest("POST", ce.endpoint, bytes.NewReader(compressedData))
+	if err != nil {
+		return err
+	}
+
+	//Add optional headers
+	for name, value := range ce.headers {
+		httpReq.Header.Set(name, value)
+	}
+
+	//Add necessary headers: https://cortexmetrics.io/docs/apis/#remote-api
+	httpReq.Header.Add("Content-Encoding", "snappy")
+	httpReq.Header.Set("Content-Type", "application/x-protobuf")
+	httpReq.Header.Set("X-Prometheus-Remote-Write-Version", "0.1.0")
+
+	//Changing context of the httpreq to global context
+	httpReq = httpReq.WithContext(ctx)
+
+	ctx, cancel := context.WithTimeout(context.Background(), ce.client.Timeout)
+	defer cancel()
+
+	httpResp, err := ce.client.Do(httpReq)
+	if err != nil {
+		return err
+	}
+
+	//Only even status codes < 400 are okay
+	if httpResp.StatusCode/100 != 2 || httpResp.StatusCode >= 400 {
```

Comment:
This condition feels unnecessarily weird... why not do something more readable:

httpResp.StatusCode != 200 || ..

Reply:
We are expecting a 2XX status code. What we were thinking is distinguishing whether an error is recoverable or not, and let the collector initiate retry based on that. We were confused on the status code and will update it to checking for 2XX.

Reply:
I'd look to narrow it down to handling specific status codes we expect to be returned by the API we're interfacing with. Services don't generally return such a broad range (all 2XX for example) of status codes unless you're interfacing with multiple endpoints and we'd generally want to know if the success status code being returned changed as it might indicate changes in functionality or something else. Those types of changes/issues would fly under the radar if we're accepting all 2XX status codes.

Reply:
Great suggestion! I think a reason we were hesitant to specify what code specifically is because theoretically the exporter could send to Cortex and other different remote write API integrated backends, and we aren't sure what the exact behavior of each backend is. We will look for a single spec and update this part.

Reply:
I looked at Cortex code and it logs an error on its side if the response code is not 202 Accepted. We will change the code to reflect the same behavior as well.